tinygrad/ops.py

```diff
@@ -40,20 +40,129 @@ def identity_element(op:BinaryOps, dt:DType): return dtypes.as_const({BinaryOps.
+  """
+  Defines the ShapeTracker for a buffer UOp ([`UOps.LOAD`](/developer/uop#tinygrad.ops.UOps.LOAD),
+  [`UOps.STORE`](/developer/uop#tinygrad.ops.UOps.STORE) or [`UOps.CONST`](/developer/uop#tinygrad.ops.UOps.CONST)).
+
+  - **`dtype`**: `None`
+  - **`src`**: `Tuple[]`
+  - **`arg`**: `ShapeTracker`
+  """
+  SWIZZLE = auto()
+  DEFINE_GLOBAL = auto()
+  DEFINE_VAR = auto()
+  DEFINE_LOCAL = auto()
+  DEFINE_ACC = auto()
+  CONST = auto()
+  SPECIAL = auto()
+  NOOP = auto()
+  GEP = auto()
   # math ops
-  CAST = auto(); BITCAST = auto(); VECTORIZE = auto() # noqa: E702
-  ALU = auto(); REDUCE = auto(); REDUCE_AXIS = auto(); WMMA = auto() # noqa: E702
+  CAST = auto()
+  """
+  - **`dtype`**: The casted scalar DType
+  - **`src`**: `Tuple[UOp]`
+  - **`arg`**: `None`
+  """
+  BITCAST = auto()
+  """
+  - **`dtype`**: The bitcasted scalar DType
+  - **`src`**: `Tuple[UOp]`
+  - **`arg`**: `None`
+  """
+  VECTORIZE = auto()
+  """
+  - **`dtype`**: The upcasted vector DType
+  - **`src`**: `Tuple[UOp]`
+  - **`arg`**: `None`
+  """
+  ALU = auto()
+  """
+  - **`dtype`**: Output DType
+  - **`src`**: `Tuple[UOp] | Tuple[UOp, UOp] | Tuple[UOp, UOp, UOp]`
+  - **`arg`**: `UnaryOps | BinaryOps | TernaryOps`
+  """
+  REDUCE = auto()
+  REDUCE_AXIS = auto()
+  """
+  - **`dtype`**: Output DType
+  - **`src`**: Input to reduce `Tuple[UOp]`
+  - **`arg`**: `(BinaryOps.ADD | BinaryOps.MUL | BinaryOps.MAX, Tuple[int, ...])`
+  """
+  WMMA = auto()
   # memory/assignment ops
-  LOAD = auto(); STORE = auto(); PHI = auto() # noqa: E702
+  LOAD = auto()
+  """
+  - **`dtype`**: Output DType
+  - **`src`**:
+    - Normal LOAD: `Tuple[UOp, UOp]`
+      - Buffer UOp [`UOps.DEFINE_GLOBAL`](/developer/uop#tinygrad.ops.UOps.DEFINE_GLOBAL).
+      - Indexing UOp, can only return `dtypes.int32`.
+    - Gated LOAD: `Tuple[UOp, UOp, UOp, UOp]`
+      - Buffer UOp [`UOps.DEFINE_GLOBAL`](/developer/uop#tinygrad.ops.UOps.DEFINE_GLOBAL).
+      - Indexing UOp, can only return `dtypes.int32`.
+      - Gate UOp, can only return `dtypes.bool`.
+      - [`UOps.CONST`](/developer/uop#tinygrad.ops.UOps.CONST) 0, 0.0 or `False`
+    - Barriered LOAD: `Tuple[UOp, UOp, UOp, UOp]`
+      - Buffer UOp [`UOps.DEFINE_LOCAL`](/developer/uop#tinygrad.ops.UOps.DEFINE_LOCAL).
+      - Indexing UOp, can only return `dtypes.int32`.
+      - Gate UOp, can only return `dtypes.bool`.
+      - Barrier UOp, [`UOps.BARRIER`](/developer/uop#tinygrad.ops.UOps.BARRIER)
+  - **`arg`**: `None`
+  """
+  STORE = auto()
+  """
+  - **`dtype`**: `None`
+  - **`src`**:
+    - Normal STORE: `Tuple[UOp, UOp, UOp]`
+      - Buffer UOp [`UOps.DEFINE_GLOBAL`](/developer/uop#tinygrad.ops.UOps.DEFINE_GLOBAL)
+        or [`UOps.DEFINE_LOCAL`](/developer/uop#tinygrad.ops.UOps.DEFINE_LOCAL).
+      - Indexing Op, can only return `dtypes.int32`.
+      - Value to store.
+    - Gated STORE: `Tuple[UOp, UOp, UOp, UOp]`
+      - Buffer UOp [`UOps.DEFINE_GLOBAL`](/developer/uop#tinygrad.ops.UOps.DEFINE_GLOBAL).
```

Comment:
Gated local allowed?

Reply:
my bad, can be local in multi reduce + pad.